Skia Sharp Mobile Blazor Bindings

lachlanwgordon opened the pull request:
Hi @Eilon and @mattleibow, I think this is just about ready to go except for the issues noted below with the SKGLCanvas, which I'm a bit stuck with. Any thought from either of you? 

Created a wrapper project to allow Skia Sharp to be used in Mobile Blazor Bindings Apps. Ideally this will be distributed as it's own Nuget.

Fixes #167 

![2020-08-15 17 02 30](https://user-images.githubusercontent.com/29908924/90307409-28359c80-df19-11ea-99ba-c647fb42410f.gif)

# API Summary
In a razor component add an `SKCanvasView` from the `Microsoft.MobileBlazorBindings.SkiaSharp` namespace.
## SKCanvasView
```
@using Microsoft.MobileBlazorBindings.SkiaSharp
...
<SKCanvasView @ref="CanvasView"
                            OnPaintSurface="PaintSurface"
                            VerticalOptions="LayoutOptions.FillAndExpand" />
```
Add a method to the component called `PaintSurface ` which takes an `SKPaintSurfaceEventArgs` parameter. The canvas is accessed using `e.Surface.Canvas`. The simplest usage is to clear the canvas and apply a color.

```
private void PaintSurface(SK.SKPaintSurfaceEventArgs e)
{
            var canvas = e.Surface.Canvas;
            canvas.Clear(SKColors.Crimson);
}
```

The PaintSurface method will automatically be called once when the view renders. If you want it to update again this is done by calling `InvalidateSurface` on the View which is accessed with `@ref=CanvasViews`

```
public SKCanvasView CanvasView { get; set; }
public void Update()
{
    CanvasView.InvalidateSurface();
}
```

## SKGLView
The SKGLView is basically the same as as the SKCanvasView but it is GPU accelerated. In high performance situations this can be useful but in my experience the SKCanvas view is much more popular. The only difference to the API is that the PaintSurface method accepts an `SKPaintGLSurfaceEventArgs` as the parameter.

There is currently an issue with the SKGLView not working with the Renderer Dispatcher on Android. See known issues below.

# Sample
See the SkiaSharpPlayground in control gallery for samples. 
Once you have a View wired up with a method to handle PaintSurface, usage is exactly the same as SkiSharp.Views.Forms which has lots of examples around. https://github.com/xamarin/xamarin-forms-samples/tree/master/SkiaSharpForms

# Documentation
Documentation PR is in progress, I'll finish it up once we've got this ready to merge.

# Known Issues
I'm having problems with using the `SKGLView` on Android.

I declare an the paint surface event in Razor Syntaxt `OnPaintSurface="PaintSurface"`. This works fine on iOS and fine on the SKCanvas but not on the GPU accelerated `SKGLView`.

When `SKGLViewHandler` tries to dispatch the event I run into all sort of weird problems
* For use on iOS DispatchEventAsync must not be inside InvokeAsync()
* Sometimes I'll get the error on Android "The current thread is not associated with the Dispatcher" which can be fixed with InvokeAsync
* If it does run, the view will only update once or twice, but some times I need to call it lots of times close together
* If I try to add break point the app crashes completely
* Sometimes the crash happens when you try to access the Canvas through the event args
* All issues go away if you add the even using the NativeControl instead of using Razor Syntax
* To see these issues run the Control Gallery on an Android, click Skia Playground, click Skia GL Paths MBB Events
* To see the GLCanvas working correctly by bypassing dispatcher run the Control Gallery, click Skia Playground, click Skia GL Native Control Events

# Solution changes
* Added MobileBlazorBindings.SkiaSharp Project
* Added samples to Control Gallery
* Added Build Configurations to allow testing on iPhone

# Questions
* Can we release this as a separate Nuget? Or will people have to use source? Or someting else?
* Given the SKGLView is not essential and it has issues on Android, should we leave it out, or include it without the broken event that can be worked around, or include it with the bug, or do you have any thoughts on fixing the threading issue?
* I didn't use the generator because it's such a small API surface(and I couldn't get it to work) but all the features are there. Does that matter, do you want the generated file included for consistency?
* @mattleibow have I missed any features?

mattleibow commented:
The Android issue is most probably this https://github.com/mono/SkiaSharp/issues/1465

I need to rework how all this works because there are a few issues on Android:

 - updates not synced to the screen refresh rate
 - updates not on the UI thread

Is it possible to allow rendering on a background thread in Blazor? UWP also allows this option. It is good for games because then the UI is not impacted by a high framer ate.

lachlanwgordon commented:
Thanks Matt.

That's interesting, I'm not quite sure if it is related to that, given it works perfectly when I add the event to the Forms control directly instead of using the MBB renderer.

e.g.
```
//This works
(CanvasView.NativeControl as SK.SKGLView).PaintSurface += SkiaSharpGLPlayground_PaintSurface;
//This doesn't
OnPaintSurface="PaintSurface"
``` 

If you're going to rework some threading stuff, that might be the cause, I think the best solution for this project is to remove the SKGLView until the Android threading issues are under control in SkiaSharp, and release this with SKCanvasView only for now. I'm happy to provide more info or test things out if I can help.

As far as rendering goes in Mobile Blazor Bindings, it is all handled by Xamarin Forms so I think all the same rules would apply. i.e. Android and iOS need to be updated on the main thread. This is complicated a little by the way MBB keeps its own UI tree and decides when to update things using the NativeComponentRenderer and Dispatcher. But I think it all boils down to Xamarin.Forms controls that must be updated on UI thread.

In WebBlazor I have found that StateHasChanged() must be called on the UI thread. StateHasChanged() is a bit like NotifyPropertyChanged but it triggers an update of the whole UI instead of just bits bound to a property. That sounded slow to me at first but they do it with clever diffing so it only changes the relevant bits, like in MVU.

mattleibow commented:
Seems fair to leave the SKGLView out for now.

Eilon commented on samples/ControlGallery/ControlGallery/Views/SkiaCanvasDemo.razor:
```diff
@@ -0,0 +1,13 @@
+﻿@using Microsoft.MobileBlazorBindings.SkiaSharp
+@page "/skiacanvas/"
```
Would it be "nice" to use hierarchical URLs? Does it work if you do like `/skia/canvas` etc. for all the pages? Otherwise some of the URLs later in the PR look like it's a keyboard mash `/skialasflahdgkahegauh` 😄

lachlanwgordon replied:
If I'm not going to use hierarchical URL, why did i bother building them? Fixed.

lachlanwgordon commented on src/Microsoft.MobileBlazorBindings.SkiaSharp/Elements/SKCanvasView.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public void InvalidateSurface()
+        {
+            (NativeControl as SK.SKCanvasView).InvalidateSurface();
```
Good point, fixed.

lachlanwgordon commented on src/Microsoft.MobileBlazorBindings.SkiaSharp/Microsoft.MobileBlazorBindings.SkiaSharp.csproj:
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
I've added all of these

lachlanwgordon commented:
I've fixed all suggestions, thanks for those.

I've removed the SKGLView, Handler and all relevant samples. I've kept the code on a separate [branch](https://github.com/lachlanwgordon/MobileBlazorBindings/tree/lachlan/SkiaSharpGLView) so we can bring it back if we work out the threading issues or if someone requests the feature and is happy to work around the limitations.

Eilon commented:
It's in!

Eilon commented:
Oh and I just added this so that it will publish NuGet packages to the feed: https://github.com/xamarin/MobileBlazorBindings/commit/c082e29c385517751d31899dd5f6782dbfd633b2
